### Titanic Survival Prediction using Machine Learning

#### Problem Statement

The objective of this project is to build a machine learning model that predicts whether a passenger survived the Titanic disaster based on passenger characteristics such as age, gender, ticket class, fare, and family size.

#### Dataset

The Titanic dataset contains demographic and travel information for passengers aboard the Titanic.

#### Tools and Technologies

- Python
- Pandas
- NumPy
- Matplotlib
- Scikit-Learn
- Jupyter Notebook

-------------------------------------------------------------------------------------------------------------
### Importing libraries

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

-------------------------------------------------------------------------------------------------------------
### Loading dataset

In [3]:
titan = pd.read_csv(r'C:\Users\chand\Desktop\AIML Learning\Titanic survival ML\Titanic-Dataset.csv')
titan.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


----------------------------------
### Data Preprocessing
- Missing value analysis
- Feature Engineering
- Encoding


In [234]:
# Missing values analysis
titan.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [235]:
# Filling the missing values
titan['Age'] = titan.Age.fillna(value = titan.Age.median())
titan['Embarked'] = titan.Embarked.fillna(value = titan.Embarked.mode()[0])
titan = titan.drop('Cabin',axis=1)
titan.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [236]:
# Feature Engineering
titan["Family_size"] = titan["SibSp"] + titan["Parch"]
titan.groupby("Family_size")["Survived"].sum()

Family_size
0     163
1      89
2      59
3      21
4       3
5       3
6       4
7       0
10      0
Name: Survived, dtype: int64

In [237]:
# Encoding
titan["Sex"] = titan["Sex"].map({"male":0,"female":1})
titan

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Family_size
0,1,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,S,1
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C,1
2,3,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,S,0
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,S,1
4,5,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,S,0
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",0,27.0,0,0,211536,13.0000,S,0
887,888,1,1,"Graham, Miss. Margaret Edith",1,19.0,0,0,112053,30.0000,S,0
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",1,28.0,1,2,W./C. 6607,23.4500,S,3
889,890,1,1,"Behr, Mr. Karl Howell",0,26.0,0,0,111369,30.0000,C,0


-----------------------------------------------
### Feature Selection and train-test split
Features:
- Pclass
- Age
- Fare
- Sex
- Family_Size

Target:
- Survived


In [238]:
X = titan[["Pclass","Age","Fare","Sex","Family_size"]]
y = titan["Survived"]
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

---------------------------------------------------------------------------
### Model training and evaluation

##### 1. Decision Tree Classifier model

In [239]:
dm = DecisionTreeClassifier(random_state=42)
dm.fit(X_train,y_train)
dm_pred = dm.predict(X_test)
dm_acc = accuracy_score(dm_pred,y_test)
print("Decision Tree Accuracy:",dm_acc)

Decision Tree Accuracy: 0.7597765363128491


------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------
### Model Comparision

To identify the most effective machine learning algorithm for predicting Titanic passenger survival, multiple classification models were trained and evaluated on the same training and testing datasets.

The models evaluated are:
- Decision Tree Classifier
- Random Forest Classifier
- Logistic Regression
- K-Nearest Neighbors (KNN)

Model performance is compared using accuracy score.

##### 2. Random Forest Classifier model

In [240]:
rf_model = RandomForestClassifier(random_state = 42)
rf_model.fit(X_train,y_train)
rf_pred = rf_model.predict(X_test)
rf_acc = accuracy_score(rf_pred,y_test)
print("Random Forest Accuracy:",rf_acc)

Random Forest Accuracy: 0.8044692737430168


##### 3. Logistic Regression model
 

In [241]:
lr_model = LogisticRegression(random_state= 42)
lr_model.fit(X_train,y_train)
lr_pred = lr_model.predict(X_test)
lr_acc = accuracy_score(y_test,lr_pred)
print("Logistic Regression Accuracy:",lr_acc)

Logistic Regression Accuracy: 0.8044692737430168


##### 4. KNN model

In [242]:
kn_model = KNeighborsClassifier()
kn_model.fit(X_train,y_train)
kn_pred = kn_model.predict(X_test)
kn_acc = accuracy_score(kn_pred,y_test)
print("KNN Accuracy:",kn_acc)

KNN Accuracy: 0.7094972067039106


---------------------------------------------------------------------------------------------------
### Comparision table

In [243]:
results = {"Decision Tree Classifier" : dm_acc,
             "Random Forest Classifier" : rf_acc,
             "Logistic Regression" : lr_acc,
             "K Neighbors Classifier" : kn_acc}

for model,score in results.items():
    print(model,":",round(score,2)*100,"%")

Decision Tree Classifier : 76.0 %
Random Forest Classifier : 80.0 %
Logistic Regression : 80.0 %
K Neighbors Classifier : 71.0 %


### Observation

Four machine learning models were evaluated on the Titanic dataset.

- Logistic Regression achieved an accuracy of 80%.
- Random Forest achieved an accuracy of 80%.
- Decision Tree achieved an accuracy of 76%.
- K-Nearest Neighbors achieved an accuracy of 71%.

Logistic Regression and Random Forest achieved the highest accuracy among the evaluated models, indicating that both models were effective for predicting passenger survival.

-----------

## Conclusion

This project explored the Titanic dataset and developed machine learning models to predict passenger survival.

The workflow included:
- Data exploration
- Missing value handling
- Feature engineering
- Model training
- Model evaluation
- Model comparison

Four classification algorithms were evaluated:
- Decision Tree Classifier
- Random Forest Classifier
- Logistic Regression
- K-Nearest Neighbors Classifier

Logistic Regression and Random Forest achieved the highest accuracy of 80%, making them the best-performing models for this dataset.

The analysis demonstrated that passenger characteristics such as gender, passenger class, age, fare, and family size contribute to survival prediction.

This project showcases an end-to-end machine learning workflow using Python, Pandas, and Scikit-Learn.